src/pysoem/pysoem.pyx

```diff
@@ -685,6 +712,8 @@ cdef class CdefSlave:
         """
         if self._ecx_contextt == NULL:
             raise UnboundLocalError()
```

Comment:
This two lines could actually be removed, right? As `__check_context_is_initialized` is doing the same?

Reply:
I'm not sure it's doing the same.  context_initialized keeps track if the Master (Context) is initialized (called open and not called close).
The previous one checks if if the slave has a reference to it's c context. I could not find anywhere where this is unset, so I don't think it will be ever triggered, but it's a different error altogether.
The master could have been deleted, and the CDefSlave instance dereferenced from the master, but this is not done.

Reply:
While checking this I've found a possible pitfall with the existing code.
CdefMaster can be dereferenced and then deleted by the python garbage collector.
This also deletes, or makes the memory invalid for _ecx_contextt, context_initialized, _ec_slave, _settings attributes of the Cython struct that are passed as a pointer to the slave with CdefMaster._get_slave(). 
So in theory an application could have a slave with non-null pointers that are invalid because the master was deleted.
I will try to replicate this and see if I can get a crash.

Reply:
Okay, I was able to replicate this but the crash is not as hard as I expected.

I created a test that replicates this:
```
import sys
from pysoem.pysoem import LargeObject
import pysoem
IFNAME = "\\Device\\NPF_{####################}" # Insert yours

def test():
    master = pysoem.Master()
    print(f"Reference count of master is {sys.getrefcount(master)}")
    # https://docs.python.org/3/library/sys.html#sys.getrefcount
    # 2 reference: variable and one temporal

    master.open(IFNAME)
    master.config_init()
    slave = master.slaves[0]

    for i in range(1000): # Check that we can do a lot of regular sdo reads
        slave.sdo_read(
            index=0x1018,
            subindex=0x01,
        )

    print(f"Reference count of master is {sys.getrefcount(master)}")
    # https://docs.python.org/3/library/sys.html#sys.getrefcount
    # Still 2 reference: variable and one temporal
    
    # Now "delete" the master. This could be any de-referencing via objects
    del master
    # Now it should have no references so it's deleted

    large_objects = []

    for i in range(1000):
        # Reading with the master destroyed. Unsafe
        # But somehow we still have an accessible slave that has references to the master that is destroyed
        slave.sdo_read(
            index=0x1018,
            subindex=0x01,
        )
        # But is successful for a while since the objects already still in the memory

        # We have to create objects that will take the memory
        large_objects.append(LargeObject())
        # And at the end something will break
```

And modified pysoem a bit to have more info:
Created the large object:
```
from libc.string cimport memset
cdef class LargeObject:
    cdef uint32_t container[0xb0]

    def __cinit__(self):
        memset(self.container, 0, sizeof(self.container))
        container_position = <uint32_t>&self.container
        size = sizeof(self.container)
        end = container_position + size

        r = range(container_position, end)
        print(f"container range {r}")
````

And added info to master:
```
CdefMaster
    .....

    def __del__(self):
        print("Master object is being destroyed")

    def __cinit__(self):
        first_element_position = <uint32_t>&self._ec_slave
        last_element_position = <uint32_t>&self.context_initialized

        cdef_master_position_range = range(first_element_position, last_element_position)
        print(f"cdef_master_position_range {cdef_master_position_range}")
```

The output is the following:
```
cdef_master_position_range range(3512352816, 3512523328)
Reference count of master is 2
Reference count of master is 2
Master object is being destroyed
container range range(3511733872, 3511734576)
container range range(3511726512, 3511727216)
container range range(3511729456, 3511730160)
container range range(3511735344, 3511736048)
container range range(3511734608, 3511735312)
container range range(3511723568, 3511724272)
container range range(3511725776, 3511726480)
container range range(3511738288, 3511738992)
container range range(3511727984, 3511728688)
container range range(3511730928, 3511731632)
container range range(3511724304, 3511725008)
container range range(3511728720, 3511729424)
container range range(3511730192, 3511730896)
container range range(3511727248, 3511727952)
container range range(3511732400, 3511733104)
container range range(3511722832, 3511723536)
container range range(3511733136, 3511733840)
container range range(3511725040, 3511725744)
container range range(3511410176, 3511410880)
container range range(3511403552, 3511404256)
container range range(3511406496, 3511407200)
container range range(3511407232, 3511407936)
container range range(3511411648, 3511412352)
container range range(3511399872, 3511400576)
container range range(3511407968, 3511408672)
container range range(3511412384, 3511413088)
container range range(3511404288, 3511404992)
container range range(3511413120, 3511413824)
container range range(3511410912, 3511411616)
container range range(3511408704, 3511409408)
container range range(3511400608, 3511401312)
container range range(3511402816, 3511403520)
container range range(3511405760, 3511406464)
container range range(3511413856, 3511414560)
container range range(3511414592, 3511415296)
container range range(3511399136, 3511399840)
container range range(3511405024, 3511405728)
container range range(3511401344, 3511402048)
container range range(3511409440, 3511410144)
container range range(3511402080, 3511402784)
container range range(3511417008, 3511417712)
container range range(3511421424, 3511422128)
container range range(3511417744, 3511418448)
container range range(3511423632, 3511424336)
container range range(3511424368, 3511425072)
container range range(3511428048, 3511428752)
container range range(3511419952, 3511420656)
container range range(3511425104, 3511425808)
container range range(3511425840, 3511426544)
container range range(3511430256, 3511430960)
container range range(3511426576, 3511427280)
container range range(3511420688, 3511421392)
container range range(3511422160, 3511422864)
container range range(3511418480, 3511419184)
container range range(3511427312, 3511428016)
container range range(3511419216, 3511419920)
container range range(3511428784, 3511429488)
container range range(3511422896, 3511423600)
container range range(3511429520, 3511430224)
container range range(3511430992, 3511431696)
container range range(3511415536, 3511416240)
container range range(3511416272, 3511416976)
container range range(3511433408, 3511434112)
container range range(3511440768, 3511441472)
container range range(3511439296, 3511440000)
container range range(3511444448, 3511445152)
container range range(3511441504, 3511442208)
container range range(3511437824, 3511438528)
container range range(3511445184, 3511445888)
container range range(3511434880, 3511435584)
container range range(3511434144, 3511434848)
container range range(3511437088, 3511437792)
container range range(3511440032, 3511440736)
container range range(3511445920, 3511446624)
container range range(3511432672, 3511433376)
container range range(3511442240, 3511442944)
container range range(3511435616, 3511436320)
container range range(3511442976, 3511443680)
container range range(3511438560, 3511439264)
container range range(3511446656, 3511447360)
container range range(3511443712, 3511444416)
container range range(3511447392, 3511448096)
container range range(3511431936, 3511432640)
container range range(3511436352, 3511437056)
container range range(3511462320, 3511463024)
container range range(3511448336, 3511449040)
container range range(3511457904, 3511458608)
container range range(3511451280, 3511451984)
container range range(3511454224, 3511454928)
container range range(3511449072, 3511449776)
container range range(3511463792, 3511464496)
container range range(3511456432, 3511457136)
container range range(3511452016, 3511452720)
container range range(3511458640, 3511459344)
container range range(3511463056, 3511463760)
container range range(3511455696, 3511456400)
container range range(3511452752, 3511453456)
container range range(3511450544, 3511451248)
container range range(3511454960, 3511455664)
container range range(3511459376, 3511460080)
container range range(3511453488, 3511454192)
container range range(3511460112, 3511460816)
container range range(3511449808, 3511450512)
container range range(3511457168, 3511457872)
container range range(3511460848, 3511461552)
container range range(3511461584, 3511462288)
container range range(3512358768, 3512359472)
container range range(3512352880, 3512353584)
FAILED
tests\context_test.py:9 (test)
def test():
        master = pysoem.Master()
    
        print(f"Reference count of master is {sys.getrefcount(master)}")
        # https://docs.python.org/3/library/sys.html#sys.getrefcount
        # 2 reference: variable and one temporal
    
        master.open(IFNAME)
    
        master.config_init()
    
        slave = master.slaves[0]
    
        for i in range(1000):
            slave.sdo_read(
                index=0x1018,
                subindex=0x01,
            )
    
        print(f"Reference count of master is {sys.getrefcount(master)}")
        # https://docs.python.org/3/library/sys.html#sys.getrefcount
        # Still 2 reference: variable and one temporal
        del master
        # Now it should have no references so it's deleted
    
        large_objects = []
    
        for i in range(1000):
            # Reading with the master destroyed. Unsafe
>           slave.sdo_read(
                index=0x1018,
                subindex=0x01,
            )

context_test.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pysoem.pysoem.WkcError

src\pysoem\pysoem.pyx:777: WkcError
```

Checking the range of memory that the master did use and the ranges that the large object was using and it's clearly overwritting the previously freed memory, causing this crash.
![image](https://github.com/user-attachments/assets/e2b9eaa8-b4e9-48db-a0d7-41c43f4aaa7a)

I have no idea why is a WkcError, but I think is just coincidental. Every time I run the script different memory locations are used and the new objects could land anywhere...

In conclusion. It is essential that the slave keeps a python reference to the master so the python garbage collector does not destroy the master while there are slaves that are referenced to it. This will also unify the exception I created into a single one